Toggle for build output

rodrmoya opened the pull request:
This introduces a service for feature switch, so that it's easy to enable/disable experimental features, and uses it to conditionally enable structured build output
> [Fixes  VSTS #701567](https://devdiv.visualstudio.com/DevDiv/_workitems/edit/701567)

Commit:
[BuildOutput] Add toggle to enable the feature

Fixes https://devdiv.visualstudio.com/DevDiv/_workitems/edit/701567

Commit:
[Core] Add FeatureSwitchService

Which provides an easy way to add feature toggles, based on an
extensible condition mechanism, so that implementations can extend
as they wish.

Therzok commented on main/src/core/MonoDevelop.Core/MonoDevelop.Core/FeatureSwitchService.cs:
```diff
@@ -0,0 +1,149 @@
+	/// <summary>
+	/// Base class for feature switch conditions.
+	/// </summary>
+	public abstract class FeatureSwitchCondition
```
Could be testable.

rodrmoya replied:
yes, adding tests in the next commit

Therzok commented on main/src/core/MonoDevelop.Core/MonoDevelop.Core/FeatureSwitchService.cs:
```diff
@@ -0,0 +1,149 @@
+	/// </summary>
+	public class AggregatedFeatureSwitchCondition : FeatureSwitchCondition
+	{
+		readonly List<FeatureSwitchCondition> conditions = new List<FeatureSwitchCondition> ();
```
Why not just an immutable array here?

rodrmoya replied:
yes, indeed, it never changes, as I don't expect the need to add conditions to `Aggregated...` once instantiated, as the idea is to use this like `InstrumentationCounter`'s API, with static fields like in https://github.com/mono/monodevelop/pull/6483/files#diff-563f5ecf5ef6e0a860fd9d77cb55312fR46

Therzok commented on main/src/core/MonoDevelop.Core/MonoDevelop.Core/FeatureSwitchService.cs:
```diff
@@ -0,0 +1,149 @@
+
+			foreach (var cond in conditions) {
+				bool r = cond.Evaluate ();
+				if (allMustPass && !r) return false;
```
Replace the two conditions with this:

```
if (allMustPass ^ r)
    return allMustPass
````

Therzok commented on main/src/core/MonoDevelop.Core/MonoDevelop.Core/FeatureSwitchService.cs:
```diff
@@ -0,0 +1,149 @@
+
+		public static FeatureSwitch RegisterFeature (string name, FeatureSwitchCondition condition)
+		{
+			var result = features [name] = new FeatureSwitch { Name = name, Condition = condition };
```
Should this be thread safe?

Therzok replied:
I'd probably use an immutabledictionary and retry loops.

```
https://devdiv.visualstudio.com/DevDiv/_wiki/wikis/DevDiv.wiki?wikiVersion=GBwikiMaster&pagePath=%2FTeam%20Pages%2FVS%20for%20Mac%2FCoding%20Guides%2FParallelism%2C%20how%20to%20avoid%20locks.%20Atomic%20operations%20and%20lock%252Dfree%20algorithms
```

Therzok commented on main/src/core/MonoDevelop.Core/MonoDevelop.Core/FeatureSwitchService.cs:
```diff
@@ -0,0 +1,149 @@
+
+		public static bool IsEnabled (string name)
+		{
+			if (features.TryGetValue (name, out FeatureSwitch fs)) {
```
Should this be thread safe?

Therzok commented on main/src/core/MonoDevelop.Core/MonoDevelop.Core/FeatureSwitchService.cs:
```diff
@@ -0,0 +1,149 @@
+		internal override bool Evaluate ()
+		{
+			if (!string.IsNullOrEmpty (nameToCheck)) {
+				return string.IsNullOrEmpty (valueToCheck) ? Environment.GetEnvironmentVariables ().Contains (nameToCheck) : Environment.GetEnvironmentVariable (nameToCheck) == valueToCheck;
```
Shouldn't this just do GetEnvironmentVariable (nameToCheck) and check `valueToCheck == null || value == valueToCheck`?

Therzok commented on main/src/core/MonoDevelop.Core/MonoDevelop.Core/FeatureSwitchService.cs:
```diff
@@ -0,0 +1,149 @@
+			return true; // FIXME: or should we assume feature is disabled if we don't know about it?
+		}
+
+		public static bool IsEnabled (FeatureSwitch fs) => fs?.IsEnabled ?? true;
```
Why do we need this?

Commit:
[Core] Cleanup implementation of FeatureSwitchService

Use an ImmutableArray for AggregatedFeatureSwitchCondition, as the list of conditions
won't change, cleanup several unneeded methods (for now at least) and add tests.

Commit:
[Tests] Add tests for BuildOutput's feature toggle

Commit:
[Core] Move FeatureSwitchService API to its own namespace
